kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -0,0 +1,386 @@
+                    edekSerde.serialize(edek, prefix);
+                    prefix.flip();
+
+                    return new KeyContext(prefix,
+                            System.nanoTime() + dekTtlNanos,
+                            maxEncryptionsPerDek,
+                            // Either we have a different Aes encryptor for each thread
+                            // or we need mutex
+                            // or we externalize the state
+                            AesGcmEncryptor.forEncrypt(new AesGcmIvGenerator(new SecureRandom()), dekPair.dek()));
+                });
+    }
+
+    @NonNull
+    @Override
+    @SuppressWarnings("java:S2445")
+    public CompletionStage<Void> encrypt(@NonNull EncryptionScheme<K> encryptionScheme,
+                                         @NonNull List<? extends Record> records,
+                                         @NonNull Receiver receiver) {
+
+        var fieldsHeader = createEncryptedFieldsHeader(encryptionScheme.recordFields());
+        return currentDekContext(encryptionScheme.kekId(), records.size()).thenAccept(keyContext -> {
+            var dekHeader = createEdekHeader(keyContext);
+            var maxValuePlaintextSize = encryptionScheme.recordFields().contains(RecordField.RECORD_VALUE)
+                    ? records.stream().mapToInt(Record::valueSize).max().orElse(-1)
+                    : -1;
+            synchronized (keyContext) {
+                var valueCiphertext = maxValuePlaintextSize >= 0 ? bufferPool.acquire(keyContext.encodedSize(maxValuePlaintextSize)) : null;
+                try {
+                    encryptRecords(encryptionScheme, keyContext, records, fieldsHeader, dekHeader, valueCiphertext, receiver);
+                }
+                finally {
+                    if (maxValuePlaintextSize >= 0) {
+                        bufferPool.release(valueCiphertext);
+                    }
+                }
+            }
+        });
+    }
+
+    private void encryptRecords(@NonNull EncryptionScheme<K> encryptionScheme,
+                                @NonNull KeyContext keyContext,
+                                @NonNull List<? extends Record> records,
+                                @NonNull RecordHeader fieldsHeader,
+                                @NonNull RecordHeader dekHeader,
+                                ByteBuffer valueCiphertext,
+                                @NonNull Receiver receiver) {
+        records.forEach(kafkaRecord -> {
+            ByteBuffer transformedValue;
+            Header[] headers = kafkaRecord.headers();
+            if (encryptionScheme.recordFields().contains(RecordField.RECORD_VALUE)) {
+                transformedValue = encryptRecordValue(keyContext, kafkaRecord, valueCiphertext);
+            }
+            else {
+                transformedValue = kafkaRecord.value();
+            }
+            if (encryptionScheme.recordFields().contains(RecordField.RECORD_HEADER_VALUES)) {
```

Comment:
Would we really want to encrypt header fields individually?  Header values are often small (boolean etc), so the repeated AES blocking overhead might be significant.  Also confidentiality/integrity is improved if the header keys are encrypted too.

I would encode the header map to a single byte array, as `org.apache.kafka.common.record.DefaultRecord#writeTo`.  I would then either encrypt the header bytes and store in a special header `kroxylicious.io/headers-encrypted`, or (better) actually have the concat the record bytes with the record headers and encrypt the whole thing once.

Reply:
> This scheme is also better as it means if Kroxylicious instances (with EE) are ever chained, we don't need to worry about headers such as `kroxylicious.io/dek` getting clobbered.

The current codes is safe in this respect: It always prepends "our" header on produce and removes them on fetch. It doesn't look at whatever existing headers are present.

But I agree with you that this isn't quite how we should do it. 

* I think using a header is the best way to guarantee that we correctly identify records that are encrypted. Anything else (I can think of) has a higher risk false positives. Such false positives manifest as decryption failing. We want decryption failing to unambiguously mean that the ciphertext bytes have been changed.
* I agree it would be better to encrypt header names as well as values.
* I'm undecided about putting them in the value. You disguise that there were some headers, I suppose. Is that something that needs obscuring?
* I'd like to consider how we serialize the metadata in some more detail. Right now I'm using two headers for that, but as I mentioned I think 1 would be cleaner. For example right now I'm encrypting all headers, but perhaps there are use cases where it's beneficial to be able to read certain specific headers without proxied access. We don't have to actually support this, but should design the serialization to accommodate it in the future.
* One way of thinking about that metadata is as instructions, to be interpreted on the fetch path, for how to restore the record. Apart from the encryption we might want to support restoring the original compression that the producer used. This would allow use to experiment with finding the best compression of the encrypted records. I don't expect the ciphertexts themselves to compress well (because they look random, which is the worst case for compression), but the metadata is a repeating pattern.

Reply:
>  I think using a header is the best way to guarantee that we correctly identify records that are encrypted.

+1 

I much prefer this idea to the heuristics. 

> I'm undecided about putting them in the value. You disguise that there were some headers, I suppose. Is that something that needs obscuring?

Isn't the less we reveal the better?

Separately, what the overhead of encrypting n header names/values  separately (i.e. n*2 encrypt operation) vs. the cost of constructing a concatenation of field name/value paisr into a single buffer and encrypting once?   I thinking before in terms of CPU and additional storage space owing to the AES blocking overhead.  I can run a little benchmark.

> perhaps there are use cases where it's beneficial to be able to read certain specific headers 

not sure.  I can imagine in the field level encryption use-case case being able exert finer control would be desirable.   My extinct for record level encryption is to keep it out of scope.


> Apart from the encryption we might want to support restoring the original compression that the producer used.

+1 -  for least surprise/symmetry 

Aside: I hadn't appreciated how badly client compression/encryption would play together until now.  I hadn't twigged that EE wouldn't be encrypting compressed bytes, but decompressed bytes. Doh!    The only benefit a producer gets from using compression when encryption is in use is a reduction in network transfer across the first hop.

Reply:
Not to join the conversation around the merits of the various options, I think its an important and valuable discussion, but I wanted to say this feels like something we should address after this PR has merged.